fix(compiler-cli): write bundled metadata via typescript compiler host

dherges opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Bundled metadata json is written directly to disk. TypeScript compiler host is skipped.

Issue Number: #21339


## What is the new behavior?

Write bundle metadata via TypeScript compiler host, thus allowing users to 'capture' bundle metadata.

Minimal Usage Example:

```ts
import * as ts from 'typescript';

  const tsCompilerHost: ts.CompilerHost = /* .. */;
  tsCompilerHost.writeFile = (
    fileName: string,
    data: string,
    writeByteOrderMark: boolean,
    onError?: (message: string) => void,
    sourceFiles?: ReadonlyArray<ts.SourceFile>
  ): void => {
    /* Customize `writeFile` callback */
  };

  const ngCompilerHost = ng.createCompilerHost({
    options: tsConfig.options,
    tsHost: tsCompilerHost
  });
```



## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```



## Other information

dherges committed:
fix: write bundled metadata via typescript compiler host

Closes #21339

dherges commented:
The CI error on Travis is:

```bash
$ ./scripts/ci/build.sh
The command "./scripts/ci/build.sh" exited with 0.

$ ./scripts/ci/test.sh
$ ng e2e --preserve-symlinks --no-webdriver-update
Version of @angular/compiler-cli needs to be 2.3.1 or greater. Current version is "6.0.0-beta.1-83a3eba".
Error: Version of @angular/compiler-cli needs to be 2.3.1 or greater. Current version is "6.0.0-beta.1-83a3eba".

(error repeats multiple times)

Suites failed:
  /home/travis/build/angular/angular/aio/content/examples/ajs-quick-reference/
  /home/travis/build/angular/angular/aio/content/examples/architecture/
  /home/travis/build/angular/angular/aio/content/examples/bootstrapping/
  /home/travis/build/angular/angular/aio/content/examples/component-interaction/
  /home/travis/build/angular/angular/aio/content/examples/dependency-injection-in-action/
  /home/travis/build/angular/angular/aio/content/examples/displaying-data/
  /home/travis/build/angular/angular/aio/content/examples/dynamic-component-loader/
  /home/travis/build/angular/angular/aio/content/examples/feature-modules/
  /home/travis/build/angular/angular/aio/content/examples/forms/
  /home/travis/build/angular/angular/aio/content/examples/i18n/
  /home/travis/build/angular/angular/aio/content/examples/lifecycle-hooks/
  /home/travis/build/angular/angular/aio/content/examples/ngmodules/
  /home/travis/build/angular/angular/aio/content/examples/providers/
  /home/travis/build/angular/angular/aio/content/examples/router/
  /home/travis/build/angular/angular/aio/content/examples/service-worker-getting-started/
  /home/travis/build/angular/angular/aio/content/examples/structural-directives/
  /home/travis/build/angular/angular/aio/content/examples/template-syntax/
  /home/travis/build/angular/angular/aio/content/examples/toh-pt1/
  /home/travis/build/angular/angular/aio/content/examples/toh-pt3/
  /home/travis/build/angular/angular/aio/content/examples/toh-pt5/
  /home/travis/build/angular/angular/aio/content/examples/user-input/
Elapsed time: 96.84 seconds
Some test suites failed
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
The command "./scripts/ci/test.sh" exited with 1.
```

Is the error due to changes on my side?

chuckjaz commented on packages/compiler-cli/src/metadata/bundle_index_host.ts:
```diff
@@ -51,7 +51,7 @@ function createSyntheticIndexHost<H extends ts.CompilerHost>(
           // If we are writing the synthetic index, write the metadata along side.
           const metadataName = fileName.replace(DTS, '.metadata.json');
-          fs.writeFileSync(metadataName, indexMetadata, {encoding: 'utf8'});
+          delegate.writeFile(metadataName, indexMetadata, writeByteOrderMark, onError, sourceFiles);
```
Can you add a test to `bundle_spec.ts`?

dherges replied:
Yes.

dherges replied:
I added the test to `program_spec.ts` as `createSyntheticIndexHost()` isn't exported and I couldn't find a way to set up test doubles for `createBundleIndexHost()`. Thus I followed the pattern of the other "should (not) emit .." tests.

dherges committed:
test(compiler-cli): verify that bundled metadata is written via ts compiler host

dherges committed:
style(compiler-cli): adjust max line length

dherges commented:
Fixed the linting issue, rebased to master, now circle ci gets a bazel error:

```
ERROR: error loading package '': Encountered error while reading extension file 'defs.bzl': no such package '@build_bazel_rules_typescript//': Error cloning repository: ssh://git@github.com/bazelbuild/rules_typescript: Algorithm negotiation fail caused by ssh://git@github.com/bazelbuild/rules_typescript: Algorithm negotiation fail caused by Algorithm negotiation fail
```

chuckjaz commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -529,6 +529,32 @@ describe('ng program', () => {
+    const oldWriteFile = host.writeFile;
+    host.writeFile = (fileName, data, writeByteOrderMark, onError, sourceFiles) => {
+      writtenFileNames.push(fileName);
+      oldWriteFile(fileName, data, writeByteOrderMark, onError, sourceFiles);
```
NIT: you probably don't need this as you never look at the disk.

petebacondarwin commented:
Closing as this is only relevant to the deprecated ViewEngine compiler and it is stale and has not been updated in over 18 months.
